core/newdelete/src/NewDelete.cxx

```diff
@@ -257,6 +241,24 @@ void *operator new(size_t size, void *vp)
 }
 #endif
 
+#if __cplusplus >= 201700L
+void *operator new(size_t size, std::align_val_t al)
+{
+   void *ptr = nullptr;
+   if (posix_memalign(&ptr, static_cast<size_t>(al), size) != 0)
```

Comment:
This makes the custom operator new with and without alignment incompatible/semantically-different.  

Note that one of the feature of the custom operator new is that when gMmallocDesc is set it will use (indirectly) mmalloc (allocation of shared memory) rather than calloc/malloc.

And since this overload will/might call implicitly depending on the circunstance, things that used to be allocated in the 'shared memory section' will (silently) no longer be.

Reply:
This is not yet used anywhere. The comments you made, although they are correct, do not give me the direction of what a correct solution would be, since the implementation of the custom placement new seems incorrect. What I care about at the moment is that experiments are waiting for a version of ROOT that compiles with C++17 support, and this (something they don't need) is blocking them. I do not think that ROOT should be replacing the allocation operators in the global namespace for all user code, and therefore my first attempt was to simply remove the "implementations" that you put in, since they don't compile. If ROOT needs to have custom operators, we can add them later into the ROOT namespace and update the code that needs them (which for now is still no code at all).

Reply:
>  do not think that ROOT should be replacing the allocation operators in the global namespace

But it already does, optionally (if you load -lNew).

> and therefore my first attempt was to simply remove the "implementations"

The point of those (intentionally broken since at the time there was no compiler to test them) implementation was as a remainder to do the right thing.   

We can do either of two things

1) overload operator new with alignment and make it fail (Fatal or exception)
2) implement properly.

> since the implementation of the custom placement new seems incorrect.

The implementation is **correct**  but the behavior is *of course* different from regular operator new.  The custom operator new without placement will allocate memory and need to *optionally* allocate this memory in the shared memory section (mmalloc).

> update the code that needs them (which for now is still no code at all).

Humm ... there exit code that really on the custom operator new (anything that uses TMapFile for example) ... It is clear that we do not *have to* support this for operator new with alignment but it needs to fail nicely (not silently) so option 1) is plausible.  Not implementing them at all is not an option because in that case **when the user** load libNew and (intentionally or not) updates its classes to now be set in a way that the compiler will use the operator new with alignment (whereas it was not doing so before), the program will now silently fail (allocating the object in regular memory rather than the shared memory as intended by the user).


So in conclusion we need the operator new without placement and with alignment to ***either***

a) throw an exception if ROOT::Internal::gMmallocDesc is true. [in which case the code you have will not do what the user expects/want/need]

b) use an implementation similar to the custom operator new (i.e. call mmalloc (or equivalent) when ROOT::Internal::gMmallocDesc is true)

Cheers,
Philippe.